doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,752 @@
+int SSL_CTX_set1_SSL_EVENT_CTX(SSL_CTX *ctx, SSL_EVENT_CTX *ectx);
+
+/*
+ * Override the SSL_EVENT_CTX to use for an inidivudal SSL object. Calling
+ * SSL_set_bio(), SSL_set0_rbio() or SSL_set0_wbio() on an SSL object will
+ * replace any existing SSL_EVENT_CTX with a new one containing those BIOs.
+ * The SSL_CTX used to create the SSL object and SSL_EVENT_CTX object must use
+ * the same SSL_METHOD.
+ */
+int SSL_set1_SSL_EVENT_CTX(SSL *ssl, SSL_EVENT_CTX *ectx);
+
+/*
+ * With this option set a single SSL_EVENT_CTX object can be safely shared and
+ * used between multiple threads.
+ */
+# define SSL_EVENT_CTX_NEW_OPTION_THREAD_SAFE       0x01
+
+/* Types of events that can occur within an SSL_EVENT_CTX */
+# define SSL_EVENT_TYPE_ERROR            0x01
+# define SSL_EVENT_TYPE_NEW_CONNECTION   0x02
+# define SSL_EVENT_TYPE_NEW_STREAM       0x04
+# define SSL_EVENT_TYPE_CONNECTION_CLOSE 0x08
+# define SSL_EVENT_TYPE_STREAM_CLOSE     0x10
+# define SSL_EVENT_TYPE_STREAM_READABLE  0x20
+# define SSL_EVENT_TYPE_STREAM_WRITEABLE 0x40
+# define SSL_EVENT_TYPE_TICK             0x80
+
+# define SSL_EVENT_TYPE_ALL_EVENTS      (SSL_EVENT_TYPE_NEW_CONNECTION \
+                                        | SSL_EVENT_TYPE_NEW_STREAM \
+                                        | SSL_EVENT_TYPE_CONNECTION_CLOSE \
+                                        | SSL_EVENT_TYPE_STREAM_CLOSE \
+                                        | SSL_EVENT_TYPE_STREAM_READABLE \
+                                        | SSL_EVENT_TYPE_STREAM_WRITEABLE \
+                                        | SSL_EVENT_TYPE_TICK)
+
+/*
+ * Create a new SSL_EVENT_CTX object for the protocol SSL_METHOD as defined on
+ * the SSL_CTX. evmask is a bitwise OR of all the events that should be reported
+ * for this SSL_EVENT_CTX. Use of an SSL_EVENT_CTX will be optional for (D)TLS
+ * but required for QUIC (since there is no way of receiving some of the events
+ * otherwise).
+ */
+SSL_EVENT_CTX *SSL_EVENT_CTX_new(SSL_CTX *ctx, uint64_t evmask, uint64_t options);
+
+int SSL_EVENT_CTX_up_ref(SSL_EVENT_CTX *ectx);
+void SSL_EVENT_CTX_free(SSL_EVENT_CTX *ectx);
+
+/*
+ * Add new default rbio's and wbio's to be used by all consumers of this
+ * SSL_EVENT_CTX. Does not need to be called for TLS. Required for DTLS if using
+ * an SSL_EVENT_CTX. Always required for QUIC.
+ */
+int SSL_EVENT_CTX_add0_rbio(SSL_EVENT_CTX *ectx, BIO *rbio);
+int SSL_EVENT_CTX_add0_wbio(SSL_EVENT_CTX *ectx, BIO *wbio);
+
+/*
+ * Process any time based actions that need to occur for the SSL object. For
+ * DTLS this is equivalent to calling DTLSv1_handle_timeout(). For TLS this does
+ * nothing (there are no time-based events for TLS). For QUIC this only has an
+ * effect on the default stream.
+ */
+int SSL_tick(SSL *s);
+
+/*
+ * Populates nexttick with the time until the soonest time based action that
+ * requires a call to SSL_tick(). For DTLS, this is equivalent to calling
```

Comment:
Not: That first sentence threw me off.  Perhaps replacing "soonest" with "next" makes it easier to parse?

Reply:
Fixed.